File: leetcode/cheapest-flights-within-k-stops/cheapest-flights-within-k-stops.py
# LeetCode #787 - Cheapest Flights Within K Stops
# https://leetcode.com/problems/cheapest-flights-within-k-stops/
# Via BFS with relaxations (compare to Dijkstra's algorithm).

class Solution:
    @staticmethod
    def findCheapestPrice(n: int,
                          flights: List[List[int]],
                          src: int,
                          dst: int,
                          k: int) -> int:
        graph = Graph(n)
        for src, dest, weight in flights:
            graph.add_edge(src, dest, weight)

        min_cost = graph.min_path_cost(src, dst, k + 1)
        return -1 if min_cost == math.inf else min_cost


class Graph:
    """A weighted directed graph."""

    __slots__ = ('_adj',)

    def __init__(self, vertex_count):
        self._adj = [[] for _ in range(vertex_count)]

    def add_edge(self, src, dest, weight):
        self._check(src)
        self._check(dest)
        self._adj[src].append((dest, weight))

    def min_path_cost(self, start, finish, max_depth):
        print(f'start={start}, finish={finish}')
        self._check(start)
        self._check(finish)

        costs = [math.inf] * self._vertex_count
        queue = collections.deque()
        costs[start] = 0
        queue.append((start, 0))

        for _ in range(max_depth):
            if not queue:
                break

            for _ in range(len(queue)):
                print(queue)
                src, src_cost = queue.popleft()

                for dest, weight in self._adj[src]:
                    dest_cost = src_cost + weight
                    if costs[dest] <= dest_cost:
                        continue

                    costs[dest] = dest_cost
                    queue.append((dest, dest_cost))

        return costs[finish]

    @property
    def _vertex_count(self):
        return len(self._adj)

    def _check(self, vertex):
        if not 0 <= vertex < self._vertex_count:
            raise ValueError('vertex out of range')

File: leetcode/cheapest-flights-within-k-stops/cheapest-flights-within-k-stops.rb
# LeetCode #787 - Cheapest Flights Within K Stops
# https://leetcode.com/problems/cheapest-flights-within-k-stops/
# Via BFS with relaxations (compare to Dijkstra's algorithm).

# @param {Integer} n
# @param {Integer[][]} flights
# @param {Integer} src
# @param {Integer} dst
# @param {Integer} k
# @return {Integer}
def find_cheapest_price(n, flights, src, dst, k)
  min_cost = build_graph(n, flights).min_path_cost(src, dst, k + 1)
  min_cost == Float::INFINITY ? -1 : min_cost
end

def build_graph(vertex_count, edges)
  graph = Graph.new(vertex_count)
  edges.each { |src, dest, weight| graph.add_edge(src, dest, weight) }
  graph
end

# A weighted directed graph.
class Graph
  def initialize(vertex_count)
    @adj = Array.new(vertex_count) { [] }
  end

  def add_edge(src, dest, weight)
    check(src)
    check(dest)
    @adj[src] << [dest, weight]
    nil
  end

  def min_path_cost(start, finish, max_depth)
    check(start)
    check(finish)

    costs = [Float::INFINITY] * vertex_count
    costs[start] = 0
    queue = [[start, 0]]

    max_depth.times do
      break if queue.empty?

      queue.size.times do
        src, src_cost = queue.shift

        @adj[src].each do |dest, weight|
          dest_cost = src_cost + weight
          next if costs[dest] <= dest_cost

          costs[dest] = dest_cost
          queue << [dest, dest_cost]
        end
      end
    end

    costs[finish]
  end

  private

  def vertex_count
    @adj.size
  end

  def check(vertex)
    raise 'vertex out of range' unless vertex.between?(0, vertex_count - 1)
  end
end
